src/functional/ffront/dialect_parser.py

```diff
@@ -0,0 +1,166 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2021, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+from __future__ import annotations
+
+import ast
+import textwrap
+import types
+from dataclasses import dataclass
+from typing import Any, ClassVar, Generic, Optional, Type, TypeVar, cast
+
+from eve.type_definitions import SourceLocation
+from functional import common
+from functional.ffront.source_utils import ClosureRefs, SourceDefinition, SymbolNames
+
+
+DialectRootT = TypeVar("DialectRootT")
+
+
+@dataclass(frozen=True, kw_only=True)
+class DialectParser(ast.NodeVisitor, Generic[DialectRootT]):
```

Comment:
Shouldn't it also be generic with respect to the syntax error type?

Reply:
No, wouldn't that be wrong actually? One could argue that the syntax error type is a property of the dialect in which case it should be derived from that type, but you should not change the syntax error type while keeping the dialect and vice-versa and thus only the dialect is generic. I made the syntax error type a class var assigning the responsibility to decide what type of error is to be thrown to the concrete parser implementation (which is the only place syntax errors should arise from).

Reply:
>   One could argue that the syntax error type is a property of the dialect in which case it should be derived from that type
In that case, the DialectError should be parameterized with the Dialect itself, like:

```python
class DialectSyntaxError(common.GTSyntaxError, Generic[DialectRootT]):
    ...

class DialectParser(ast.NodeVisitor, Generic[DialectRootT]):
    ...
    syntax_error_cls: ClassVar[Type[DialectSyntaxError[DialectRootT]]]
```
But anyway this is not supported in Python because `ClassVar`s cannot hold `TypeVar`s, so I guess the current solution is the best one.